File: java/src/main/java/org/hyperledger/fabric/client/ProposalBuilder.java
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hyperledger.fabric.client;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import com.google.protobuf.ByteString;
import org.hyperledger.fabric.protos.common.ChannelHeader;
import org.hyperledger.fabric.protos.common.Header;
import org.hyperledger.fabric.protos.common.HeaderType;
import org.hyperledger.fabric.protos.gateway.ProposedTransaction;
import org.hyperledger.fabric.protos.peer.ChaincodeHeaderExtension;
import org.hyperledger.fabric.protos.peer.ChaincodeID;
import org.hyperledger.fabric.protos.peer.ChaincodeInput;
import org.hyperledger.fabric.protos.peer.ChaincodeInvocationSpec;
import org.hyperledger.fabric.protos.peer.ChaincodeProposalPayload;
import org.hyperledger.fabric.protos.peer.ChaincodeSpec;
import org.hyperledger.fabric.protos.peer.SignedProposal;

final class ProposalBuilder implements Proposal.Builder {
    private final GatewayClient client;
    private final SigningIdentity signingIdentity;
    private final String channelName;
    private final ChaincodeID chaincodeId;
    private final ChaincodeInput.Builder inputBuilder = ChaincodeInput.newBuilder();
    private final ChaincodeProposalPayload.Builder payloadBuilder = ChaincodeProposalPayload.newBuilder();
    private Set<String> endorsingOrgs = Collections.emptySet();

    ProposalBuilder(final GatewayClient client, final SigningIdentity signingIdentity,
                    final String channelName, final String chaincodeName, final String transactionName) {
        this.client = client;
        this.signingIdentity = signingIdentity;
        this.channelName = channelName;
        this.chaincodeId = ChaincodeID.newBuilder()
                .setName(chaincodeName)
                .build();

        inputBuilder.addArgs(ByteString.copyFrom(transactionName, StandardCharsets.UTF_8));
    }

    @Override
    public ProposalBuilder addArguments(final byte[]... args) {
        for (byte[] arg : args) {
            inputBuilder.addArgs(ByteString.copyFrom(arg));
        }
        return this;
    }

    @Override
    public ProposalBuilder addArguments(final String... args) {
        for (String arg : args) {
            inputBuilder.addArgs(ByteString.copyFrom(arg, StandardCharsets.UTF_8));
        }
        return this;
    }

    @Override
    public ProposalBuilder putAllTransient(final Map<String, byte[]> transientData) {
        transientData.forEach(this::putTransient);
        return this;
    }

    @Override
    public ProposalBuilder putTransient(final String key, final byte[] value) {
        payloadBuilder.putTransientMap(key, ByteString.copyFrom(value));
        return this;
    }

    @Override
    public ProposalBuilder putTransient(final String key, final String value) {
        payloadBuilder.putTransientMap(key, ByteString.copyFromUtf8(value));
        return this;
    }

    @Override
    public ProposalBuilder setEndorsingOrganizations(final String... mspids) {
        this.endorsingOrgs = new HashSet<>(Arrays.asList(mspids));
        return this;
    }

    @Override
    public Proposal build() {
        return new ProposalImpl(client, signingIdentity, channelName, newProposedTransaction());
    }

    private ProposedTransaction newProposedTransaction() {
        TransactionContext context = new TransactionContext(signingIdentity);

        return ProposedTransaction.newBuilder()
                .setProposal(newSignedProposal(context))
                .setTransactionId(context.getTransactionId())
                .addAllEndorsingOrganizations(endorsingOrgs)
                .build();
    }

    private SignedProposal newSignedProposal(final TransactionContext context) {
        return SignedProposal.newBuilder()
                .setProposalBytes(newProposal(context).toByteString())
                .build();
    }

    private org.hyperledger.fabric.protos.peer.Proposal newProposal(final TransactionContext context) {
        return org.hyperledger.fabric.protos.peer.Proposal.newBuilder()
                .setHeader(newHeader(context).toByteString())
                .setPayload(newChaincodeProposalPayload().toByteString())
                .build();
    }

    private Header newHeader(final TransactionContext context) {
        return Header.newBuilder()
                .setChannelHeader(newChannelHeader(context).toByteString())
                .setSignatureHeader(context.getSignatureHeader().toByteString())
                .build();
    }

    private ChannelHeader newChannelHeader(final TransactionContext context) {
        return ChannelHeader.newBuilder()
                .setType(HeaderType.ENDORSER_TRANSACTION.getNumber())
                .setTxId(context.getTransactionId())
                .setTimestamp(GatewayUtils.getCurrentTimestamp())
                .setChannelId(channelName)
                .setExtension(newChaincodeHeaderExtension().toByteString())
                .setEpoch(0)
                .build();
    }

    private ChaincodeHeaderExtension newChaincodeHeaderExtension() {
        return ChaincodeHeaderExtension.newBuilder()
                .setChaincodeId(chaincodeId)
                .build();
    }

    private ChaincodeProposalPayload newChaincodeProposalPayload() {
        return payloadBuilder
                .setInput(newChaincodeInvocationSpec().toByteString())
                .build();
    }

    private ChaincodeInvocationSpec newChaincodeInvocationSpec() {
        return ChaincodeInvocationSpec.newBuilder()
                .setChaincodeSpec(newChaincodeSpec())
                .build();
    }

    private ChaincodeSpec newChaincodeSpec() {
        return ChaincodeSpec.newBuilder()
                .setChaincodeId(chaincodeId)
                .setInput(inputBuilder.build())
                .build();
    }
}

File: node/src/proposalbuilder.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { common, gateway, peer } from '@hyperledger/fabric-protos';
import { Timestamp } from 'google-protobuf/google/protobuf/timestamp_pb';
import { GatewayClient } from './client';
import { Proposal, ProposalImpl } from './proposal';
import { SigningIdentity } from './signingidentity';
import { TransactionContext } from './transactioncontext';

/**
 * Options used when evaluating or endorsing a transaction proposal.
 */
export interface ProposalOptions {
    /**
     * Arguments passed to the transaction function.
     */
    arguments?: (string | Uint8Array)[];

    /**
     * Private data passed to the transaction function but not recorded on the ledger.
     */
    transientData?: Record<string, string | Uint8Array>;

    /**
     * Specifies the set of organizations that will attempt to endorse the proposal.
     * No other organizations' peers will be sent this proposal.
     * This is usually used in conjunction with transientData for private data scenarios.
     */
    endorsingOrganizations?: string[];
}

export interface ProposalBuilderOptions extends ProposalOptions {
    client: GatewayClient;
    signingIdentity: SigningIdentity;
    channelName: string;
    chaincodeName: string;
    transactionName: string;
}

export class ProposalBuilder {
    readonly #options: Readonly<ProposalBuilderOptions>;
    readonly #transactionContext: TransactionContext;

    constructor(options: Readonly<ProposalBuilderOptions>) {
        this.#options = options;
        this.#transactionContext = new TransactionContext(options.signingIdentity);
    }

    build(): Proposal {
        return new ProposalImpl({
            client: this.#options.client,
            signingIdentity: this.#options.signingIdentity,
            channelName: this.#options.channelName,
            proposedTransaction: this.#newProposedTransaction(),
        });
    }

    #newProposedTransaction(): gateway.ProposedTransaction {
        const result = new gateway.ProposedTransaction();
        result.setProposal(this.#newSignedProposal());
        result.setTransactionId(this.#transactionContext.getTransactionId());
        if (this.#options.endorsingOrganizations) {
            result.setEndorsingOrganizationsList(this.#options.endorsingOrganizations);
        }
        return result;
    }

    #newSignedProposal(): peer.SignedProposal {
        const result = new peer.SignedProposal();
        result.setProposalBytes(this.#newProposal().serializeBinary());
        return result;
    }

    #newProposal(): peer.Proposal {
        const result = new peer.Proposal();
        result.setHeader(this.#newHeader().serializeBinary());
        result.setPayload(this.#newChaincodeProposalPayload().serializeBinary());
        return result;
    }

    #newHeader(): common.Header {
        const result = new common.Header();
        result.setChannelHeader(this.#newChannelHeader().serializeBinary());
        result.setSignatureHeader(this.#transactionContext.getSignatureHeader().serializeBinary());
        return result;
    }

    #newChannelHeader(): common.ChannelHeader {
        const result = new common.ChannelHeader();
        result.setType(common.HeaderType.ENDORSER_TRANSACTION);
        result.setTxId(this.#transactionContext.getTransactionId());
        result.setTimestamp(Timestamp.fromDate(new Date()));
        result.setChannelId(this.#options.channelName);
        result.setExtension$(this.#newChaincodeHeaderExtension().serializeBinary());
        result.setEpoch(0);
        return result;
    }

    #newChaincodeHeaderExtension(): peer.ChaincodeHeaderExtension {
        const result = new peer.ChaincodeHeaderExtension();
        result.setChaincodeId(this.#newChaincodeID());
        return result;
    }

    #newChaincodeID(): peer.ChaincodeID {
        const result = new peer.ChaincodeID();
        result.setName(this.#options.chaincodeName);
        return result;
    }

    #newChaincodeProposalPayload(): peer.ChaincodeProposalPayload {
        const result = new peer.ChaincodeProposalPayload();
        result.setInput(this.#newChaincodeInvocationSpec().serializeBinary());
        const transientMap = result.getTransientmapMap();
        for (const [key, value] of Object.entries(this.#getTransientData())) {
            transientMap.set(key, value);
        }
        return result;
    }

    #newChaincodeInvocationSpec(): peer.ChaincodeInvocationSpec {
        const result = new peer.ChaincodeInvocationSpec();
        result.setChaincodeSpec(this.#newChaincodeSpec());
        return result;
    }

    #newChaincodeSpec(): peer.ChaincodeSpec {
        const result = new peer.ChaincodeSpec();
        result.setType(peer.ChaincodeSpec.Type.NODE);
        result.setChaincodeId(this.#newChaincodeID());
        result.setInput(this.#newChaincodeInput());
        return result;
    }

    #newChaincodeInput(): peer.ChaincodeInput {
        const result = new peer.ChaincodeInput();
        result.setArgsList(this.#getArgsAsBytes());
        return result;
    }

    #getArgsAsBytes(): Uint8Array[] {
        return Array.of(this.#options.transactionName, ...(this.#options.arguments ?? [])).map(asBytes);
    }

    #getTransientData(): Record<string, Uint8Array> {
        const result: Record<string, Uint8Array> = {};

        for (const [key, value] of Object.entries(this.#options.transientData ?? {})) {
            result[key] = asBytes(value);
        }

        return result;
    }
}

function asBytes(value: string | Uint8Array): Uint8Array {
    return typeof value === 'string' ? Buffer.from(value) : value;
}
